### PR TITLE
REST API structure using Slim framework

### PR DESCRIPTION
**Pull request by ArthurHoaro**
**Warning**: PHP requirement upgraded to PHP 5.5 - #599 
- REST API routes are handle by Slim.
- Every API controller go through ApiMiddleware which handles security.
- First service implemented `/info`, for example/tests purpose.

This is a first step regarding #655. It's actually pretty easy to use. :)

---

All requests must include a valid JWT token in their header.
You can use this script to generate a token (valid during 9min):

``` php
<?php

function generateJwtToken($secret)
    {
        $header = base64_encode('{
            "typ": "JWT",
            "alg": "HS512"
        }');
        $payload = base64_encode('{
            "iat": '. time() .'
        }');
        $signature = hash_hmac('sha512', $header .'.'. $payload , $secret);
        return $header .'.'. $payload .'.'. $signature;
    }

echo generateJwtToken('jaimeLePate');
```

---

New settings (which will be updated automatically):
- `api.enabled`: true/false.
- `api.secret`: API Password.
- `dev.debug`: JSON formatting and gives more details on errors.

### CODE REVIEW

**Comment by virtualtam** [comment]
#666, the PR of the beast? \m/

File: composer.json [hunk 1]
```diff
@@ -10,13 +10,21 @@
     },
     "keywords": ["bookmark", "link", "share", "web"],
     "require": {
-        "php": ">=5.3.4",
-        "shaarli/netscape-bookmark-parser": "1.*"
+        "php": ">=5.6",
```
Comment (virtualtam) [inline composer.json]: Slims requires PHP >= 5.5.9, so we could keep 5.5.x in the dependency array (Travis CI as well)
- https://github.com/slimphp/Slim/blob/3.x/.travis.yml
- https://packagist.org/packages/slim/slim#3.5.0

File: composer.json [hunk 2]
```diff
@@ -10,13 +10,21 @@
...
+        "shaarli/netscape-bookmark-parser": "1.*",
+        "slim/slim": "^3.0"
```
Comment (virtualtam) [inline composer.json]: why not using 3.5.x?

File: index.php [hunk 3]
```diff
@@ -2204,4 +2208,33 @@ function resizeImage($filepath)
...
+    $conf->get('resource.datastore'),
+    isLoggedIn(),
+    $conf->get('privacy.hide_public_links'),
+    $conf->get('redirector.url'),
+    $conf->get('redirector.encode_url')
+);
+
+$container = new \Slim\Container();
+$container['conf'] = $conf;
+//$container['db'] = $linkDb;
```
Comment (virtualtam) [inline index.php]: commented code

File: application/api/ApiMiddleware.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace Api;
```
Comment (virtualtam) [inline application/api/ApiMiddleware.php]: shouldn't we define Shaarli-specific namespaces, e.g.

```php
namespace Shaarli\Api;
```

File: application/api/ApiMiddleware.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+use \Slim\Container;
```
Comment (virtualtam) [inline application/api/ApiMiddleware.php]: what's the syntactic difference between `\Slim\Container` and `Slim\Http\Response`?

File: tpl/install.html [hunk 6]
```diff
@@ -14,6 +14,18 @@
...
                 <input type="checkbox" name="updateCheck" id="updateCheck" checked="checked"><label for="updateCheck">&nbsp;Notify me when a new release is ready</label></td>
             </tr>
+            <tr>
+                <td valign="top">
+                    <b>API:</b>
+                </td>
+                <td>
+                    <input type="checkbox" name="enableApi" id="enableApi" checked="checked">
+                    <label for="enableApi">
+                        &nbsp;Enable Shaarli's API.
```
Comment (virtualtam) [inline tpl/install.html]: +REST

File: tpl/configure.html [hunk 7]
```diff
@@ -80,6 +80,20 @@
           <label for="updateCheck">&nbsp;Notify me when a new release is ready</label>
         </td>
       </tr>
+      <tr>
+        <td valign="top"><b>Enable API</b></td>
```
Comment (virtualtam) [inline tpl/configure.html]: +REST

File: .htaccess [hunk 8]
```diff
@@ -0,0 +1,4 @@
+RewriteEngine On
+RewriteCond %{REQUEST_FILENAME} !-f
+RewriteCond %{REQUEST_FILENAME} !-d
+RewriteRule ^ index.php [QSA,L]
```
Comment (virtualtam) [inline .htaccess]: newline :smiling_imp:

File: application/api/exceptions/ApiAuthorizationException.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+     * to much security information.
+     *
+     * @param $message string the exception message.
+     */
+    public function setMessage($message)
+    {
+        $original = $this->debug === true ? ': '. $this->getMessage() : '';
+        $this->message = $message . $original;
+    }
+}
```
Comment (virtualtam) [inline application/api/exceptions/ApiAuthorizationException.php]: newline :smiling_imp:

File: tests/api/controllers/InfoTest.php [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+        $this->assertEquals(7, $data['global_counter']);
+        $this->assertEquals(2, $data['private_counter']);
+        $this->assertEquals($title, $data['settings']['title']);
+        $this->assertEquals($headerLink, $data['settings']['header_link']);
+        $this->assertEquals($timezone, $data['settings']['timezone']);
+        $this->assertEquals($enabledPlugins, $data['settings']['enabled_plugins']);
+        $this->assertEquals($defaultPrivateLinks, $data['settings']['default_private_links']);
+    }
+}
```
Comment (virtualtam) [inline tests/api/controllers/InfoTest.php]: newline :smiling_imp:

File: .htaccess [hunk 11]
```diff
@@ -0,0 +1,4 @@
+RewriteEngine On
```
Comment (virtualtam) [inline .htaccess]: #668 raised to update Nginx config

File: application/api/ApiMiddleware.php (hunk 4 above)
Comment (ArthurHoaro) [inline application/api/ApiMiddleware.php]: Yes, I wanted to change that and I forgot!

File: application/api/ApiMiddleware.php (hunk 5 above)
Comment (ArthurHoaro) [inline application/api/ApiMiddleware.php]: Probably difference between manual and IDE. I'll remove the trailing \

File: composer.json (hunk 2 above)
Comment (ArthurHoaro) [inline composer.json]: I used their docs, and `^3.0` actually means `>=3.0 && <4.0` ([see](https://getcomposer.org/doc/articles/versions.md)). I don't know, should we set a fixed version number?

**Comment by ArthurHoaro** [comment]
Thanks for the review, updated. 
What about the milestone, regarding PHP compatibility? Do we just put it in 0.9.0 (and release 0.8.1 soon)?

**Comment by virtualtam** [comment]
Yep, sounds good! We can tackle some minor issues in 0.8.1 and move major/unstarted works to 0.9.0.

I'll focus on #351 to cleanup `LinkDB` stuff :p

**Comment by ArthurHoaro** [comment]
Rebased. OK to merge?

File: .travis.yml [hunk 12]
```diff
@@ -7,8 +7,6 @@ php:
   - 7.0
   - 5.6
   - 5.5
-  - 5.4
```
Comment (virtualtam) [inline .travis.yml]: => dedicated commit as per #599

File: CHANGELOG.md [hunk 13]
```diff
@@ -7,8 +7,12 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [v0.9.0](https://github.com/shaarli/Shaarli/releases/tag/v0.9.0) - UNPUBLISHED
 
+**WARNING**: Shaarli now requires PHP 5.5+.
```
Comment (virtualtam) [inline CHANGELOG.md]: => dedicated commit as per #599

File: composer.json [hunk 14]
```diff
@@ -10,14 +10,22 @@
     },
     "keywords": ["bookmark", "link", "share", "web"],
     "require": {
-        "php": ">=5.3.4",
+        "php": ">=5.5",
```
Comment (virtualtam) [inline composer.json]: => dedicated commit as per #599

File: tests/api/ApiUtilsTest.php [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+    }
+    
+    /**
+     * Generate a valid JWT token.
+     *
+     * @param string $secret API secret used to generate the signature.
+     *
+     * @return string Generated token.
+     */
+    public static function generateValidJwtToken($secret)
```
Comment (virtualtam) [inline tests/api/ApiUtilsTest.php]: should be either out of the test class, or grouped at the beginning along `setUp()` and `tearDown()` stuff

File: .htaccess (hunk 11 above)
Comment (virtualtam) [inline .htaccess]: => note in the wiki for Apache HTTPD users to enable `mod_rewrite`

File: application/api/ApiMiddleware.php [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    public function __construct($container)
+    {
+        $this->container = $container;
+        $this->conf = $this->container->get('conf');
+        $this->setLinkDb($this->conf);
+    }
+
+    /**
+     * Example middleware invokable class
```
Comment (virtualtam) [inline application/api/ApiMiddleware.php]: if it's in use, then it's not an example anymore :D
https://www.slimframework.com/docs/concepts/middleware.html

File: .htaccess (hunk 11 above)
Comment (ArthurHoaro) [inline .htaccess]: I've updated the Apache [wiki](https://github.com/shaarli/Shaarli/wiki/Server-configuration#htaccess). nginx section will have to be updated too.

**Pull request merged**